# Reading Data From Different Sources


In [1]:
import pandas as pd
from io import StringIO # Importing StringIO to read JSON data from a string
Data = '{"employee_name": "James", "email": "james@gmail.com", "job_profile": [{"title1":"Team Lead", "title2":"Sr. Developer"}]}'
df=pd.read_json(StringIO(Data))
df

,employee_name,email,job_profile
0,James,james@gmail.com,"{'title1': 'Team Lead', 'title2': 'Sr. Develop..."


In [4]:
df.to_json(orient='records') # Converting the DataFrame to JSON format with 'records' orientation 
# What is records orientation in pandas to_json method?
# The 'records' orientation in the pandas to_json method formats the JSON output as a list of dictionaries, 
# where each dictionary represents a row in the DataFrame. 
# Each key in the dictionary corresponds to a column name, and the value is the data for that column in that row. 
# This format is particularly useful for representing tabular data in JSON format, making it easy to read and work with in various applications.

'[{"employee_name":"James","email":"james@gmail.com","job_profile":{"title1":"Team Lead","title2":"Sr. Developer"}}]'

In [ ]:
pd.read_csv('https://raw.githubusercontent.com/datasciencedojo/datasets/master/titanic.csv') 
# Reading a CSV file from a URL and creating a DataFrame

,PassengerId,Survived,Pclass,Name,Sex,Age,SibSp,Parch,Ticket,Fare,Cabin,Embarked
0,1,0,3,"Braund, Mr. Owen Harris",male,22.0,1,0,A/5 21171,7.2500,NaN,S
1,2,1,1,"Cumings, Mrs. John Bradley (Florence Briggs Th...",female,38.0,1,0,PC 17599,71.2833,C85,C
2,3,1,3,"Heikkinen, Miss. Laina",female,26.0,0,0,STON/O2. 3101282,7.9250,NaN,S
3,4,1,1,"Futrelle, Mrs. Jacques Heath (Lily May Peel)",female,35.0,1,0,113803,53.1000,C123,S
4,5,0,3,"Allen, Mr. William Henry",male,35.0,0,0,373450,8.0500,NaN,S
...,...,...,...,...,...,...,...,...,...,...,...,...
886,887,0,2,"Montvila, Rev. Juozas",male,27.0,0,0,211536,13.0000,NaN,S
887,888,1,1,"Graham, Miss. Margaret Edith",female,19.0,0,0,112053,30.0000,B42,S
888,889,0,3,"Johnston, Miss. Catherine Helen ""Carrie""",female,NaN,1,2,W./C. 6607,23.4500,NaN,S
889,890,1,1,"Behr, Mr. Karl Howell",male,26.0,0,0,111369,30.0000,C148,C


In [ ]:
pd.read_csv('https://raw.githubusercontent.com/datasciencedojo/datasets/master/titanic.csv', header=None) 
# Reading a CSV file from a URL and creating a DataFrame, with header=None to indicate that the first row does not contain column names. 
# When header=None is used, pandas will automatically assign integer column names starting from 0 to the DataFrame.   

,0,1,2,3,4,5,6,7,8,9,10,11
0,PassengerId,Survived,Pclass,Name,Sex,Age,SibSp,Parch,Ticket,Fare,Cabin,Embarked
1,1,0,3,"Braund, Mr. Owen Harris",male,22,1,0,A/5 21171,7.25,NaN,S
2,2,1,1,"Cumings, Mrs. John Bradley (Florence Briggs Th...",female,38,1,0,PC 17599,71.2833,C85,C
3,3,1,3,"Heikkinen, Miss. Laina",female,26,0,0,STON/O2. 3101282,7.925,NaN,S
4,4,1,1,"Futrelle, Mrs. Jacques Heath (Lily May Peel)",female,35,1,0,113803,53.1,C123,S
...,...,...,...,...,...,...,...,...,...,...,...,...
887,887,0,2,"Montvila, Rev. Juozas",male,27,0,0,211536,13,NaN,S
888,888,1,1,"Graham, Miss. Margaret Edith",female,19,0,0,112053,30,B42,S
889,889,0,3,"Johnston, Miss. Catherine Helen ""Carrie""",female,NaN,1,2,W./C. 6607,23.45,NaN,S
890,890,1,1,"Behr, Mr. Karl Howell",male,26,0,0,111369,30,C148,C


In [ ]:
pd.read_csv('https://raw.githubusercontent.com/datasciencedojo/datasets/master/titanic.csv').to_csv('titanic.csv', index=False)
# Saving the DataFrame to a CSV file named 'titanic.csv' without including the index column in the output file.

In [18]:
url="https://www.fdic.gov/resources/resolutions/bank-failures/failed-bank-list/"

df = pd.read_html(url)


In [ ]:
df # why is it printed as a list?
# The output is printed as a list because the pd.read_html() function returns a list of DataFrames. 
# Each DataFrame in the list corresponds to a table found in the HTML content of the specified URL.
# In this case, the URL contains multiple tables, and pd.read_html() has extracted all of them, resulting in a list of DataFrames. 
# If you want to access a specific table, you can index the list, for example, df[0] to access the first table.

[                                            Bank Name                City  \
 0                   Metropolitan Capital Bank & Trust             Chicago   
 1                        The Santa Anna National Bank          Santa Anna   
 2                                Pulaski Savings Bank             Chicago   
 3                  The First National Bank of Lindsay             Lindsay   
 4               Republic First Bank dba Republic Bank        Philadelphia   
 5                                       Citizens Bank            Sac City   
 6                            Heartland Tri-State Bank             Elkhart   
 7                                 First Republic Bank       San Francisco   
 8                                      Signature Bank            New York   
 9                                 Silicon Valley Bank         Santa Clara   
 10                                  Almena State Bank              Almena   
 11                         First City Bank of Florida   Fort Wa

In [20]:
df[0]

,Bank Name,City,State,Cert,Acquiring Institution,Closing Date,Fund Sort ascending
0,Metropolitan Capital Bank & Trust,Chicago,Illinois,57488,First Independence Bank,"January 30, 2026",10550
1,The Santa Anna National Bank,Santa Anna,Texas,5520,Coleman County State Bank,"June 27, 2025",10549
2,Pulaski Savings Bank,Chicago,Illinois,28611,Millennium Bank,"January 17, 2025",10548
3,The First National Bank of Lindsay,Lindsay,Oklahoma,4134,First Bank & Trust Co.,"October 18, 2024",10547
4,Republic First Bank dba Republic Bank,Philadelphia,Pennsylvania,27332,"Fulton Bank, National Association","April 26, 2024",10546
5,Citizens Bank,Sac City,Iowa,8758,Iowa Trust & Savings Bank,"November 3, 2023",10545
6,Heartland Tri-State Bank,Elkhart,Kansas,25851,"Dream First Bank, N.A.","July 28, 2023",10544
7,First Republic Bank,San Francisco,California,59017,"JPMorgan Chase Bank, N.A.","May 1, 2023",10543
8,Signature Bank,New York,New York,57053,"Flagstar Bank, N.A.","March 12, 2023",10540
9,Silicon Valley Bank,Santa Clara,California,24735,First–Citizens Bank & Trust Company,"March 10, 2023",10539


In [22]:
url="https://en.wikipedia.org/wiki/Mobile_country_code"
pd.read_html(url,match="Country",header=0)[0]
# Reading tables from a Wikipedia page, matching the table that contains the header "Country" and using the first row as the header for the DataFrame.

HTTPError: HTTP Error 403: Forbidden

pd.read_html(url) is hitting Wikipedia directly with pandas’ default request, and Wikipedia is returning 403 Forbidden because that request looks too “bot-like”.

In [23]:
pd.read_html(
    url,
    match="Country",
    header=0,
    storage_options={"User-Agent": "Mozilla/5.0"},
)[0]
# Reading tables from a Wikipedia page, matching the table that contains the header "Country" and using the first row as the header for the DataFrame, 
# while also specifying a custom User-Agent in the storage options to avoid potential issues with web scraping.

,Mobile country code,Country,ISO 3166,Mobile network codes,National MNC authority,Remarks
0,289,A Abkhazia,GE-AB,List of mobile network codes in Abkhazia,NaN,MCC is not listed by ITU
1,412,Afghanistan,AF,List of mobile network codes in Afghanistan,NaN,NaN
2,276,Albania,AL,List of mobile network codes in Albania,NaN,NaN
3,603,Algeria,DZ,List of mobile network codes in Algeria,NaN,NaN
4,544,American Samoa (United States of America),AS,List of mobile network codes in American Samoa,NaN,NaN
...,...,...,...,...,...,...
247,452,Vietnam,VN,List of mobile network codes in the Vietnam,NaN,NaN
248,543,W Wallis and Futuna,WF,List of mobile network codes in Wallis and Futuna,NaN,NaN
249,421,Y Yemen,YE,List of mobile network codes in the Yemen,NaN,NaN
250,645,Z Zambia,ZM,List of mobile network codes in Zambia,NaN,NaN


In [26]:
df_excel = pd.read_excel('data.xlsx')

In [27]:
df_excel.to_pickle('data.pkl') # Saving the DataFrame to a pickle file named 'data.pkl' for later use.

In [28]:
pd.read_pickle('data.pkl') # Reading the DataFrame from the pickle file 'data.pkl' and loading it back into memory.

,Name,Age
0,Krish,32
1,Jack,34
2,John,31
